AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/DefaultKeyStorageInterface.dfy

```diff
@@ -708,6 +773,522 @@ module DefaultKeyStorageInterface {
+            storageOperation:="QueryForVersions",
+            ddbOperation:="Query",
+            identifier:=input.Identifier,
+            tableName:=ddbTableName));
+
+      /** Process sensical DDB Response */
+      var lastKeyBlob: seq<Types.UInt.uint8> := [];
+      var lastKeyEmpty: bool :=
+        // It is not clear if SDKs/Smithy-Dafny/Custom Implementations of Storage will respond with None or an Empty Map.
+        ddbResponse.LastEvaluatedKey.None? || (ddbResponse.LastEvaluatedKey.Some? && |ddbResponse.LastEvaluatedKey.value| == 0);
+
+      if (!lastKeyEmpty) {
+        lastKeyBlob :- lastEvaluatedKeyToBlob(ddbResponse.LastEvaluatedKey.value);
+      }
+      if (ddbResponse.Items.None? || ( |ddbResponse.Items.value| == 0) ) {
+        return Success(Types.QueryForVersionsOutput(
+                         exclusiveStartKey := lastKeyBlob,
+                         items := []
+                       ));
+      }
+
+      var items: seq<Types.EncryptedHierarchicalKey> :- Seq.MapWithResult(
+        (item: DDB.AttributeMap)
+        =>
+          var key :- encryptedHierarchicalKeyFromItem(item, logicalKeyStoreName, input.Identifier);
+          :- Need(
+               key.Type.HierarchicalSymmetricVersion?,
+               Types.KeyStorageException(
+                 message:="Unexpected item returned by DDB. TableName: " + ddbTableName + "\tBranch Key ID: " + input.Identifier
+               ));
+          Success(key),
+        ddbResponse.Items.value
+      );
+
+      return Success(
+          Types.QueryForVersionsOutput(
+            exclusiveStartKey := lastKeyBlob,
+            items := items
+          ));
+    }
+
+    function method blobToExclusiveStartKey(
+      blob: seq<Types.UInt.uint8>,
+      identifier: string
+    ): (output: Result<DDB.Key,Types.Error>)
+    {
+      // From DDB's Docs:
+      // https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html#API_Query_RequestSyntax
+      // > The primary key of the first item that this
+      // > operation will evaluate. Use the value that was returned for LastEvaluatedKey in the previous operation.
+      // > The data type for ExclusiveStartKey must be String, Number, or Binary. No set data types are allowed.
+      // From that, we can infer that Partition Key is just going to be the Identifier.
+      // Thus, we only need to store the Type value.
+      // This will be the full "branch:version:<uuidv4>"
+      var versionStr :- UTF8.Decode(blob)
+                        .MapFailure(
+                          eString
+                          =>
+                            Types.KeyStorageException(
+                              message:="Could not UTF8 Decode Exclusive Start Key. " + eString));
+      :- Need(
+           // I elected to require len > 15, rather than len == 51, in case we or someone else ever uses not-UUIDv4 for version.
+           && 15 < |versionStr| && versionStr[0 .. 15] == "branch:version:",
```

Comment:
Well... one day... we might switch to UUID v7... 
or at least support UUID v7...

It solves some problems.
I made Ryan be strict about UUID v4 strings over in the Hierarchical Keyring's 
EDK logic.

I did that so Storage would not care.

To a certain degree,
it would be wonderful is Storage did not care about the Hierarchical Keyring Version.

i.e: This storage layer works with MOST changes we can envision to Branch Keys.